setup cmp window, item.abbr maxwidth and minwidth

Song-Tianxiang opened the pull request:
Sometimes, cmp completion window, item.abbr too long. this commit can restrict item.abbr width. Maybe the default maxwidth and minwidth can be adjust.

before:
<img width="1766" alt="Screenshot 2024-10-10 at 8 25 52 PM" src="https://github.com/user-attachments/assets/b3824153-7e6c-4bb0-b2b8-28eaa19c14f6">

after:
<img width="1766" alt="Screenshot 2024-10-10 at 8 25 17 PM" src="https://github.com/user-attachments/assets/397dcded-da9f-41f1-ba76-a473908f5034">

Song-Tianxiang committed:
setup cmp window, iterm.abbr maxwidth and minwidth

siduck commented:
is there not a better way to do this      :/  , i dont like such computations in cmp config as it'd render on every keystroke!

lucario387 commented:
It's looking way too similar to this https://github.com/hrsh7th/nvim-cmp/issues/980#issuecomment-1121773499

Song-Tianxiang commented:
> is there not a better way to do this :/ , i dont like such computations in cmp config as it'd render on every keystroke!有没有更好的方法来做到这一点：/，我不喜欢 cmp 配置中的此类计算，因为它会在每次击键时呈现！

I think there's two ways to solve:
1. set a constant number but less flexible
2. compute once and save it, but since vim.o.columns would updated when window resiezed, maybe we need a `autocmd VimResized`

Song-Tianxiang commented:
> It's looking way too similar to this [hrsh7th/nvim-cmp#980 (comment)](https://github.com/hrsh7th/nvim-cmp/issues/980#issuecomment-1121773499)它看起来和这个太相似了 [hrsh7th/nvim-cmp#980 (comment)](https://github.com/hrsh7th/nvim-cmp/issues/980#issuecomment-1121773499)

shamelessly stole  from there

Song-Tianxiang committed:
set abbr_maxwidth to fixed number

lucario387 commented on lua/nvchad/cmp/init.lua:
```diff
@@ -26,7 +26,7 @@ local M = {
       -- item.abbr maxwidth and minwidth
       local ellipsis_char = '…'
-      local abbr_maxwidth = math.floor(vim.o.columns / 2)
+      local abbr_maxwidth = 29
```
`au VimResized` with a variable storing the width will be better (if it's possible)

Song-Tianxiang committed:
Merge branch 'NvChad:v3.0' into patch-1

Song-Tianxiang committed:
setup abbr.maxwidth `au VimResized`

exposed an option ` cmp.format_abbr` to user:
```
    cmp        = {
        format_abbr = {
            maxwidth = 20,
            minwidth = 20,
        }
    }
```

Song-Tianxiang commented:
I have added an option `cmp.format_abbr`.
the current `nvconfig.ui.cmp` is 
```lua
    cmp = {
        icons_left = false, -- only for non-atom styles!
        lspkind_text = true,
        style = "default", -- default/flat_light/flat_dark/atom/atom_colored
        format_colors = {
            tailwind = false, -- will work for css lsp too
            icon = "󱓻",
        },
        format_abbr = {
            maxwidth = 20,
            minwidth = 10,
        }
    },
```

How about make it look like:
```lua
    cmp = {
        icons_left = false, -- only for non-atom styles!
        lspkind_text = true,
        style = "default", -- default/flat_light/flat_dark/atom/atom_colored
        format = {
            colors = {
                tailwind = false, -- will work for css lsp too
                icon = "󱓻",
            },
            abbr = {
                maxwidth = 20,
                minwidth = 10,
            }
        },
    }
```
Its' more clean but not back compatiable (we need to use `cmp.format_colors or cmp.format.colors`). 

There's maybe other `cmp.format.[something]` in the future though.

siduck commented:
ugh i dislike such large & complex changes :/      

cant we just use string:sub?

Song-Tianxiang committed:
use string.sub & deal with maxwidth < minwidth

I can not find a better way...

Song-Tianxiang commented:
I don’t think I can come up with a better solution...

Song-Tianxiang committed:
Simplify

Song-Tianxiang commented:
Does adding this autocmd make the code seem too complex to you? The entire cost of this commit is that for each completion item, it needs to determine if it is too long, which is an unavoidable cost. The cost of the autocmd is almost negligible.

Song-Tianxiang committed:
`bit` always available

siduck commented:
dont close this yet, i dont think the autocmd is much of useful, I want the user to keep a fixed maxwidth. not depending on the nvim's window size.

Song-Tianxiang committed:
abbr_maxwidth with fixed number

siduck committed:
doc: abbr_maxwidth for cmp

siduck committed:
Merge branch 'v3.0' into patch-1

siduck commented:
@Song-Tianxiang 

```lua
vim.api.nvim_create_autocmd("VimResized", {
  callback = function()
    -- do whatever with vim.o.columns
    require("nvconfig").ui.cmp.abbr_maxwidth = 10 
  end,
})
```
